File: merging/js/quick_sort.js
function partition(list, low, high) {
  let pivot = list[low];
  let i = low;
  let j = high;
  do {
    // this while loop should stop when list[i] > pivot
    do {
      i++;
    } while (pivot >= list[i]);

    do {
      j--;
    } while (list[j] > pivot);

    if (i < j) {
      let temp = list[i];
      list[i] = list[j];
      list[j] = temp;
    }
  } while (i < j);
  let temp = list[low];
  list[low] = list[j];
  list[j] = temp;

  return j;
}

function quickSort(list, low, high) {
  let j;
  if (low < high) {
    j = partition(list, low, high);
    quickSort(list, low, j);
    quickSort(list, j + 1, high);
  }
  return list;
}

console.log(quickSort([7, 1, 9, 3, 2], 0, 5));

File: merging/cpp/quick_sort.cpp
#include <iostream>
using namespace std;

void swap(int *x, int *y)
{
  int temp = *x;
  *x = *y;
  *y = temp;
};

int partition(int A[], int low, int high)
{
  int pivot = A[low];
  int i = low;
  int j = high;

  do
  {
    do
    {
      i++;
    } while (A[i] <= pivot);
    do
    {
      j--;
    } while (A[j] > pivot);
    if (i < j)
    {
      swap(&A[i], &A[j]);
    }

  } while (i < j);
  swap(&A[low], &A[j]);
  return j;
}

void quicksort(int A[], int low, int high)
{
  int j;
  if (low < high)
  {
    j = partition(A, low, high);
    quicksort(A, low, j);
    quicksort(A, j + 1, high);
  }
}

int main()
{
  int A[] = {11, 13, 7, 12, 16, 9, 24, 5, 10, 3}, n = 10;
  quicksort(A, 0, n);

  for (int i = 0; i < 10; i++)
  {
    cout << A[i] << endl;
  }
  return 0;
}